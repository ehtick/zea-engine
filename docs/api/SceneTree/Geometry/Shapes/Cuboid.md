<a name="Cuboid"></a>

### Cuboid 
A class for generating a cuboid geometry.

**Parameters**
* **x(`NumberParameter`):** Length of the line cuboid along the `X` axis
* **y(`NumberParameter`):** Length of the line cuboid along the `Y` axis
* **z(`NumberParameter`):** Length of the line cuboid along the `Z` axis
* **BaseZAtZero(`NumberParameter`):** Property to start or not `Z` axis from position `0.


**Extends**: <code>Mesh</code>  

* [Cuboid ⇐ <code>Mesh</code>](#Cuboid)
    * [new Cuboid(x, y, z, baseZAtZero)](#new-Cuboid)
    * [setSize(x, y, z)](#setSize)
    * [setBaseSize(x, y)](#setBaseSize)
    * [toJSON() ⇒ <code>object</code>](#toJSON)

<a name="new_Cuboid_new"></a>

### new Cuboid
Create a cuboid.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| x | <code>number</code> | <code>1</code> | The length of the cuboid along the X axis. |
| y | <code>number</code> | <code>1</code> | The length of the cuboid along the Y axis. |
| z | <code>number</code> | <code>1</code> | The length of the cuboid along the Z axis. |
| baseZAtZero | <code>boolean</code> | <code>false</code> | The baseZAtZero value. |

<a name="Cuboid+setSize"></a>

### setSize
Setter for the size of the cuboid.



| Param | Type | Description |
| --- | --- | --- |
| x | <code>number</code> | The length of the edges along the X axis. |
| y | <code>number</code> | The length of the edges along the Y axis. |
| z | <code>number</code> | The length of the edges along the Z axis. |

<a name="Cuboid+setBaseSize"></a>

### setBaseSize
Setter for the base size of the cuboid.



| Param | Type | Description |
| --- | --- | --- |
| x | <code>number</code> | The length of the edges along the X axis. |
| y | <code>number</code> | The length of the edges along the Y axis. |

<a name="Cuboid+toJSON"></a>

### toJSON
The toJSON method encodes this type as a json object for persistence.


**Returns**: <code>object</code> - - Returns the json object.  
