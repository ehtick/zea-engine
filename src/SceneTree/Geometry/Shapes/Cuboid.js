import { Vec2 } from '../../../Math/Vec2';
import { Vec3 } from '../../../Math/Vec3';
import { Mesh } from '../Mesh.js';
import { BooleanParameter, NumberParameter } from '../../Parameters';
import { sgFactory } from '../../SGFactory.js';

/** Class representing a cuboid.
 * @extends Mesh
 */
class Cuboid extends Mesh {
  /**
   * Create a cuboid.
   * @param {number} x - The x value.
   * @param {number} y - The y value.
   * @param {number} z - The z value.
   * @param {boolean} baseZAtZero - The baseZAtZero value.
   */
  constructor(x = 1.0, y = 1.0, z = 1.0, baseZAtZero = false) {
    super();

    if (isNaN(x) || isNaN(y) || isNaN(z)) throw new Error('Invalid geom args');

    this.__xParam = this.addParameter(new NumberParameter('x', x));
    this.__yParam = this.addParameter(new NumberParameter('y', y));
    this.__zParam = this.addParameter(new NumberParameter('z', z));
    this.__baseZAtZeroParam = this.addParameter(
      new BooleanParameter('baseZAtZero', baseZAtZero)
    );

    this.setFaceCounts([0, 6]);
    this.setFaceVertexIndices(0, 0, 1, 2, 3);
    this.setFaceVertexIndices(1, 7, 6, 5, 4);

    this.setFaceVertexIndices(2, 1, 0, 4, 5);
    this.setFaceVertexIndices(3, 3, 2, 6, 7);

    this.setFaceVertexIndices(4, 0, 3, 7, 4);
    this.setFaceVertexIndices(5, 2, 1, 5, 6);
    this.setNumVertices(8);
    this.addVertexAttribute('texCoords', Vec2);
    this.addVertexAttribute('normals', Vec3);
    this.__rebuild();

    const resize = () => {
      this.__resize();
    };
    this.__xParam.valueChanged.connect(resize);
    this.__yParam.valueChanged.connect(resize);
    this.__zParam.valueChanged.connect(resize);
    this.__baseZAtZeroParam.valueChanged.connect(resize);
  }

  /**
   * The setSize method.
   * @param {number} x - The x param.
   * @param {number} y - The y param.
   * @param {number} z - The z param.
   */
  setSize(x, y, z) {
    this.__xParam.setValue(x);
    this.__yParam.setValue(y);
    this.__zParam.setValue(z);
  }

  /**
   * The setBaseSize method.
   * @param {number} x - The x param.
   * @param {number} y - The y param.
   */
  setBaseSize(x, y) {
    this.__xParam.setValue(x);
    this.__yParam.setValue(y);
  }

  /**
   * The __rebuild method.
   * @private
   */
  __rebuild() {
    const normals = this.getVertexAttribute('normals');
    for (let i = 0; i < 6; i++) {
      let normal;
      switch (i) {
        case 0:
          normal = new Vec3(0, 0, 1);
          break;
        case 1:
          normal = new Vec3(0, 0, -1);
          break;
        case 2:
          normal = new Vec3(1, 0, 0);
          break;
        case 3:
          normal = new Vec3(-1, 0, 0);
          break;
        case 4:
          normal = new Vec3(0, 1, 0);
          break;
        case 5:
          normal = new Vec3(0, -1, 0);
          break;
      }
      normals.setFaceVertexValue(i, 0, normal);
      normals.setFaceVertexValue(i, 1, normal);
      normals.setFaceVertexValue(i, 2, normal);
      normals.setFaceVertexValue(i, 3, normal);
    }
    const texCoords = this.getVertexAttribute('texCoords');
    for (let i = 0; i < 6; i++) {
      texCoords.setFaceVertexValue(i, 0, new Vec2(0, 0));
      texCoords.setFaceVertexValue(i, 1, new Vec2(1, 0));
      texCoords.setFaceVertexValue(i, 2, new Vec2(1, 1));
      texCoords.setFaceVertexValue(i, 3, new Vec2(0, 1));
    }
    this.__resize();
  }

  /**
   * The __resize method.
   * @param {any} mode - The mode param.
   * @private
   */
  __resize(mode) {
    const x = this.__xParam.getValue();
    const y = this.__yParam.getValue();
    const z = this.__zParam.getValue();
    const baseZAtZero = this.__baseZAtZeroParam.getValue();
    let zoff = 0.5;
    if (baseZAtZero) zoff = 1.0;
    this.getVertex(0).set(0.5 * x, -0.5 * y, zoff * z);
    this.getVertex(1).set(0.5 * x, 0.5 * y, zoff * z);
    this.getVertex(2).set(-0.5 * x, 0.5 * y, zoff * z);
    this.getVertex(3).set(-0.5 * x, -0.5 * y, zoff * z);

    zoff = -0.5;
    if (baseZAtZero) zoff = 0.0;
    this.getVertex(4).set(0.5 * x, -0.5 * y, zoff * z);
    this.getVertex(5).set(0.5 * x, 0.5 * y, zoff * z);
    this.getVertex(6).set(-0.5 * x, 0.5 * y, zoff * z);
    this.getVertex(7).set(-0.5 * x, -0.5 * y, zoff * z);

    this.setBoundingBoxDirty();
    this.geomDataChanged.emit();
  }

  /**
   * The toJSON method.
   * @return {any} - The return value.
   */
  toJSON() {
    const json = super.toJSON();
    json['x'] = this.__x;
    json['y'] = this.__y;
    json['z'] = this.__z;
    return json;
  }
}

sgFactory.registerClass('Cuboid', Cuboid);

export { Cuboid };
// export default Cuboid;
