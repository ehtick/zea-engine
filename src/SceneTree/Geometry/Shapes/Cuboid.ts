import { Vec2 } from '../../../Math/Vec2'
import { BooleanParameter } from '../../../SceneTree/Parameters/BooleanParameter'
import { NumberParameter } from '../../../SceneTree/Parameters/NumberParameter'
import { Vec3 } from '../../../Math/Vec3'
import { ProceduralMesh } from './ProceduralMesh'
import { Registry } from '../../../Registry'
import { Vec3Attribute } from '../Vec3Attribute'
import { Vec2Attribute } from '../Vec2Attribute'

/**
 * A class for generating a cuboid geometry.
 *
 * **Parameters**
 * * **x(`NumberParameter`):** Length of the line cuboid along the `X` axis
 * * **y(`NumberParameter`):** Length of the line cuboid along the `Y` axis
 * * **z(`NumberParameter`):** Length of the line cuboid along the `Z` axis
 * * **BaseZAtZero(`NumberParameter`):** Property to start or not `Z` axis from position `0.
 *
 * @extends {ProceduralMesh}
 */
class Cuboid extends ProceduralMesh {
  protected __baseZAtZeroParam: BooleanParameter
  protected __xParam: NumberParameter
  protected __yParam: NumberParameter
  protected __zParam: NumberParameter

  /**
   * Create a cuboid.
   * @param {number} x - The length of the cuboid along the X axis.
   * @param {number} y - The length of the cuboid along the Y axis.
   * @param {number} z - The length of the cuboid along the Z axis.
   * @param {boolean} baseZAtZero - The baseZAtZero value.
   */
  constructor(x = 1.0, y = 1.0, z = 1.0, baseZAtZero = false) {
    super()

    if (isNaN(x) || isNaN(y) || isNaN(z)) throw new Error('Invalid geom args')

    this.__xParam = this.addParameter(new NumberParameter('X', x)) as NumberParameter
    this.__yParam = this.addParameter(new NumberParameter('Y', y)) as NumberParameter
    this.__zParam = this.addParameter(new NumberParameter('Z', z)) as NumberParameter
    this.__baseZAtZeroParam = this.addParameter(new BooleanParameter('BaseZAtZero', baseZAtZero)) as BooleanParameter

    this.setFaceCounts([0, 6])
    this.setFaceVertexIndices(0, [0, 1, 2, 3])
    this.setFaceVertexIndices(1, [7, 6, 5, 4])

    this.setFaceVertexIndices(2, [1, 0, 4, 5])
    this.setFaceVertexIndices(3, [3, 2, 6, 7])

    this.setFaceVertexIndices(4, [0, 3, 7, 4])
    this.setFaceVertexIndices(5, [2, 1, 5, 6])
    this.setNumVertices(8)
    this.addVertexAttribute('texCoords', Vec2)
    this.addVertexAttribute('normals', Vec3)
  }

  /**
   * Setter for the size of the cuboid.
   *
   * @param {number} x - The length of the edges along the X axis.
   * @param {number} y - The length of the edges along the Y axis.
   * @param {number} z - The length of the edges along the Z axis.
   */
  setSize(x: number, y: number, z: number): void {
    this.__xParam.setValue(x)
    this.__yParam.setValue(y)
    this.__zParam.setValue(z)
  }

  /**
   * Setter for the base size of the cuboid.
   *
   * @param {number} x - The length of the edges along the X axis.
   * @param {number} y - The length of the edges along the Y axis.
   */
  setBaseSize(x: number, y: number): void {
    this.__xParam.setValue(x)
    this.__yParam.setValue(y)
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const normals: Vec3Attribute =  <Vec3Attribute>this.getVertexAttribute('normals')
    if (normals) {
      for (let i = 0; i < 6; i++) {
        let normal:  Vec3
        switch (i) {
          case 0:
            normal = new Vec3(0, 0, 1)
            break
          case 1:
            normal = new Vec3(0, 0, -1)
            break
          case 2:
            normal = new Vec3(1, 0, 0)
            break
          case 3:
            normal = new Vec3(-1, 0, 0)
            break
          case 4:
            normal = new Vec3(0, -1, 0)
            break
          //case 5:
          default:
            normal = new Vec3(0, 1, 0)
            break
        }
        normals.setFaceVertexValue(i, 0, normal)
        normals.setFaceVertexValue(i, 1, normal)
        normals.setFaceVertexValue(i, 2, normal)
        normals.setFaceVertexValue(i, 3, normal)
      }
    }
    const texCoords = <Vec2Attribute>this.getVertexAttribute('texCoords')
    if (texCoords) {
      for (let i = 0; i < 6; i++) {
        texCoords.setFaceVertexValue(i, 0, new Vec2(0, 0))
        texCoords.setFaceVertexValue(i, 1, new Vec2(1, 0))
        texCoords.setFaceVertexValue(i, 2, new Vec2(1, 1))
        texCoords.setFaceVertexValue(i, 3, new Vec2(0, 1))
      }
    }
    this.resize()
  }

  /**
   * The resize method.
   * @private
   */
  resize(): void {
    const x = this.__xParam.getValue() || 1.0
    const y = this.__yParam.getValue() || 1.0
    const z = this.__zParam.getValue() || 1.0
    const baseZAtZero = this.__baseZAtZeroParam.getValue()
    let zoff = 0.5
    const positions =  <Vec3Attribute>this.getVertexAttribute('positions')
    if (baseZAtZero) zoff = 1.0
    if (!positions) return
    positions.getValueRef(0).set(0.5 * x, -0.5 * y, zoff * z)
    positions.getValueRef(1).set(0.5 * x, 0.5 * y, zoff * z)
    positions.getValueRef(2).set(-0.5 * x, 0.5 * y, zoff * z)
    positions.getValueRef(3).set(-0.5 * x, -0.5 * y, zoff * z)

    zoff = -0.5
    if (baseZAtZero) zoff = 0.0
    positions.getValueRef(4).set(0.5 * x, -0.5 * y, zoff * z)
    positions.getValueRef(5).set(0.5 * x, 0.5 * y, zoff * z)
    positions.getValueRef(6).set(-0.5 * x, 0.5 * y, zoff * z)
    positions.getValueRef(7).set(-0.5 * x, -0.5 * y, zoff * z)
  }
}

Registry.register('Cuboid', Cuboid)

export { Cuboid }
