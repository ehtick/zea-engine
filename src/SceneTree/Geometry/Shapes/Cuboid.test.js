import { Cuboid } from './Cuboid'

describe('Cuboid', () => {
  it('tests default parameters', () => {
    const cuboid = new Cuboid()

    expect(cuboid.getParameter('X').getValue()).toBe(1)
    expect(cuboid.getParameter('Y').getValue()).toBe(1)
    expect(cuboid.getParameter('Z').getValue()).toBe(1)
    expect(cuboid.getParameter('BaseZAtZero').getValue()).toBe(false)
  })

  it('updates parameters', () => {
    const cuboid = new Cuboid()
    cuboid.getParameter('X').setValue(3)
    cuboid.getParameter('Y').setValue(4)
    cuboid.getParameter('Z').setValue(5)
    cuboid.getParameter('BaseZAtZero').setValue(true)

    expect(cuboid.getParameter('X').getValue()).toBe(3)
    expect(cuboid.getParameter('Y').getValue()).toBe(4)
    expect(cuboid.getParameter('Z').getValue()).toBe(5)
    expect(cuboid.getParameter('BaseZAtZero').getValue()).toBe(true)
  })
})
