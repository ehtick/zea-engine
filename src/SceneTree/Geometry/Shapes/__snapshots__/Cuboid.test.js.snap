// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cuboid saves to JSON (serialization). 1`] = `
Object {
  "faceCounts": Array [
    0,
    6,
  ],
  "faceVertexIndices": Array [
    0,
    1,
    2,
    3,
    7,
    6,
    5,
    4,
    1,
    0,
    4,
    5,
    3,
    2,
    6,
    7,
    0,
    3,
    7,
    4,
    2,
    1,
    5,
    6,
  ],
  "numVertices": 8,
  "params": Object {
    "BaseZAtZero": Object {
      "value": false,
    },
    "X": Object {
      "value": 4,
    },
    "Y": Object {
      "value": 6,
    },
    "Z": Object {
      "value": 4,
    },
  },
  "type": "Cuboid",
  "vertexAttributes": Object {
    "normals": Object {
      "data": Array [
        0,
        0,
        1,
        0,
        0,
        1,
        0,
        0,
        1,
        0,
        0,
        1,
        0,
        0,
        -1,
        0,
        0,
        -1,
        0,
        0,
        -1,
        0,
        0,
        -1,
      ],
      "dataType": "Vec3",
      "defaultValue": 1.7976931348623157e+308,
      "length": 8,
      "splitValues": Array [
        Object {
          "x": 1,
          "y": 0,
          "z": 0,
        },
        Object {
          "x": 1,
          "y": 0,
          "z": 0,
        },
        Object {
          "x": 1,
          "y": 0,
          "z": 0,
        },
        Object {
          "x": 1,
          "y": 0,
          "z": 0,
        },
        Object {
          "x": -1,
          "y": 0,
          "z": 0,
        },
        Object {
          "x": -1,
          "y": 0,
          "z": 0,
        },
        Object {
          "x": -1,
          "y": 0,
          "z": 0,
        },
        Object {
          "x": -1,
          "y": 0,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": 1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": 1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": 1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": 1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
      ],
      "splits": Object {
        "0": Object {
          "2": 1,
          "4": 8,
        },
        "1": Object {
          "2": 0,
          "5": 13,
        },
        "2": Object {
          "3": 5,
          "5": 12,
        },
        "3": Object {
          "3": 4,
          "4": 9,
        },
        "4": Object {
          "2": 2,
          "4": 11,
        },
        "5": Object {
          "2": 3,
          "5": 14,
        },
        "6": Object {
          "3": 6,
          "5": 15,
        },
        "7": Object {
          "3": 7,
          "4": 10,
        },
      },
    },
    "positions": Object {
      "data": Array [
        2,
        -3,
        2,
        2,
        3,
        2,
        -2,
        3,
        2,
        -2,
        -3,
        2,
        2,
        -3,
        -2,
        2,
        3,
        -2,
        -2,
        3,
        -2,
        -2,
        -3,
        -2,
      ],
      "dataType": "Vec3",
      "defaultValue": 0,
      "length": 8,
      "splitValues": Array [],
      "splits": Object {},
    },
    "texCoords": Object {
      "data": Array [
        0,
        0,
        1,
        0,
        1,
        1,
        0,
        1,
        0,
        1,
        1,
        1,
        1,
        0,
        0,
        0,
      ],
      "dataType": "Vec2",
      "defaultValue": 1.7976931348623157e+308,
      "length": 8,
      "splitValues": Array [
        Object {
          "x": 0,
          "y": 0,
        },
        Object {
          "x": 1,
          "y": 0,
        },
        Object {
          "x": 1,
          "y": 1,
        },
        Object {
          "x": 0,
          "y": 1,
        },
        Object {
          "x": 0,
          "y": 0,
        },
        Object {
          "x": 1,
          "y": 0,
        },
        Object {
          "x": 1,
          "y": 1,
        },
        Object {
          "x": 0,
          "y": 1,
        },
        Object {
          "x": 1,
          "y": 0,
        },
        Object {
          "x": 1,
          "y": 1,
        },
        Object {
          "x": 0,
          "y": 0,
        },
        Object {
          "x": 0,
          "y": 1,
        },
      ],
      "splits": Object {
        "0": Object {
          "2": 1,
        },
        "1": Object {
          "2": 0,
        },
        "2": Object {
          "3": 5,
          "5": 10,
        },
        "3": Object {
          "3": 4,
          "4": 8,
        },
        "4": Object {
          "2": 2,
        },
        "5": Object {
          "2": 3,
        },
        "6": Object {
          "3": 6,
          "5": 11,
        },
        "7": Object {
          "3": 7,
          "4": 9,
        },
      },
    },
  },
}
`;
